Update DB2OnCloudSetup.md #893 (#892)
IBM/FHIR: docs/src/pages/guides/DB2OnCloudSetup.md

layout: post
title:  Setup for IBM Db2 on Cloud
description: Setup for IBM Db2 on Cloud
date:   2020-04-03 09:59:05 -0400
permalink: /FHIR/guides/DB2OnCloudSetup
---

## Using Db2 for persistence

This document guides a developer or administrator through the steps necessary to setup and configure [IBM Db2 on IBM Cloud](https://cloud.ibm.com/catalog/services/db2) as a persistence layer for the IBM FHIR Server.

### **Create a Db2 instance**

1. Log in to your IBM Cloud account [Link](https://cloud.ibm.com/).

1. Click `Create resource`.

1. Choose `Db2 (transactional)`.

1. Select the Pricing Plan: 
    - The IBM FHIR Server recommends the Flex plan for a production workload.
    - The IBM FHIR Server recommends the Flex plan for development, however, the Lite plan is possible for the development and evaluation. 

>       - **Note** The Lite plan has a limit of 5 concurrent connections, and the IBM FHIR Server `fhirProxyDataSource` needs to be updated to avoid failures. One should update the server.xml connectionManager with maxPoolSize - `<connectionManager maxPoolSize="5"/>.  Further, your instance may not be enabled with IAM enabled, and you may use the `Service Credentials` that are created to connect to configure the datasource. This is only acceptable with **development** and **experimentation**. 

Reviewer: ```suggestion
      - **Note** The Lite plan has a limit of 5 concurrent connections, and the IBM FHIR Server `fhirProxyDataSource` needs to be updated to avoid failures. One should update the server.xml connectionManager with maxPoolSize - `<connectionManager maxPoolSize="5"/>`.  Further, your instance may not be enabled with IAM enabled, and you may use the `Service Credentials` that are created to connect to configure the datasource. This is only acceptable with **development** and **experimentation**. 
```

>       - **Note** The Lite plan has a limit of 5 concurrent connections, and the IBM FHIR Server `fhirProxyDataSource` needs to be updated to avoid failures. One should update the server.xml connectionManager with maxPoolSize - `<connectionManager maxPoolSize="5"/>`.  Further, your instance may not be enabled with IAM enabled, and you may use the `Service Credentials` that are created to connect to configure the datasource. This is only acceptable with **development** and **experimentation**. 

Reviewer: did you actually try this?  last time i tried it with the 'Lite' plan, it ran out of room.  this is why we have https://github.com/IBM/FHIR/issues/819

Author: Yeah - I wanted to check on this.  I think took what we had there. and move it around . Let me move this to the appendix with additional details.

Author: Updated (in coming commit)

      
1. Create `Create`

Your instance is now creating or created.

### **Scale the Instance**

> If you chose the Flex plan, you may want to scale the instance after it has been created (e.g. 4 cores, 16GB). The instance can be scaled more than once, so it doesn't matter if you don't get the sizing right first time. Please note, the scaling of the instance requires a service restart.

Reviewer: The cpu/memory can be scaled up or down, but the storage can only go up (no scaling down storage).  Probably worth mentioning.

Author: done

        "password": "bluadmin-password-very-secret",
        "https_url": "https://dashdb-txn-flex-************.services.dal.bluemix.net:8443",
        "port": 50000,
        "ssldsn": "DATABASE=BLUDB;HOSTNAME=dashdb-txn-flex-************.services.dal.bluemix.net;PORT=50001;PROTOCOL=TCPIP;UID=bluadmin;PWD=bluadmin-password-very-secret;Security=SSL;",
        "host": "dashdb-txn-flex-************.services.dal.bluemix.net",
        "jdbcurl": "jdbc:db2://dashdb-txn-flex-************.services.dal.bluemix.net:50000/BLUDB",
        "uri": "db2://bluadmin:bluadmin-password-very-secret@dashdb-txn-flex-************.services.dal.bluemix.net:50000/BLUDB",
        "db": "BLUDB",
        "dsn": "DATABASE=BLUDB;HOSTNAME=dashdb-txn-flex-************.services.dal.bluemix.net;PORT=50000;PROTOCOL=TCPIP;UID=bluadmin;PWD=bluadmin-password-very-secret;",
        "username": "bluadmin",
        "ssljdbcurl": "jdbc:db2://dashdb-txn-flex-************.services.dal.bluemix.net:50001/BLUDB:sslConnection=true;"
    }
    ```

4. Save these details for later, as these properties are needed to deploy and manage the IBM FHIR Server schema.

> This **BLUADMIN** user is used to deploy the schema objects (tables, indexes, stored procedures etc) only. 

Reviewer: ```suggestion
For improved security, the **BLUADMIN** user should only be used to deploy the schema objects (tables, indexes, stored procedures etc) and administer the database, NOT for connecting from the fhir-server application. 
```

### **Add the FHIRSERVER User and API Key**

Following the least-privilege principle, the IBM FHIR server itself does not use **BLUADMIN**. The IBM FHIR server uses an API Key associated with an [IAM Service Id](https://cloud.ibm.com/iam/serviceids). The IAM Service Id is mapped to a Db2 user which is granted explicit privileges to the tables and stored procedures.

The IBM FHIR Server uses the access flow: 

> 1. Read API Key from the fhir-server-config.json

Reviewer: worth mentioning the tenantKey here as well?

Author: done

> 2. Connect to IAM which maps the ServiceId to a Db2 User

Reviewer: We don't do this.  We use the apikey to connect to db2.  Db2 validates the apikey itself.  This is why we set `"securityMechanism": 15` in the connection properties in `fhir-server-config.json`

Author: let me clarify ... the blob of text depicting the flow wasn't well described.

Author: Expanded/Clarified

            ```
        1. You see `0` if not IAM enabled. 

1. Click Add. A panel opens on the right-hand side.

1. Select Add IBMid User at the top.

    - User ID: FHIRSERVER

    - IBMid: paste the service id `fhir-server-id` from the Service Id created previously.

    - Note (1): The page forces the value to lower-case, so `ServiceId` becomes `serviceid`. Don't be alarmed, it still works.

    - Note (2): Do NOT select Administrator. One should follow the least-privelege principal for the FHIRSERVER user.

1. Click `Create`.

You are now able to connect to the database as the FHIRSERVER user using only the API key created above.

### **Testing the Connection**

The [Db2 driver](https://repo1.maven.org/maven2/com/ibm/db2/jcc/11.5.0.0/jcc-11.5.0.0.jar) is able to execute a connectivity test to check the configuration of the combo of API-key/Service-Id/Db2-User-Id.

1. Copy the command to your code editor

    ``` bash
    java -cp /path/to/db2jcc4.jar com.ibm.db2.jcc.DB2Jcc  -url "jdbc:db2://<DB2-HOSTNAME>:50001/BLUDB:apiKey=<API-KEY>;securityMechanism=15;sslConnection=true;sslTrustStoreLocation=/path/to/truststore.jks;sslTrustStorePassword=<TRUSTSTORE-PASSWORD>;"
    ```

>     - Note: Don't forget the trailing `;` in the URL. Some of the documented examples don't include it, but it is required in order for the connection to work, although this may be fixed in a future driver release. This only affects this test URL, not the actual FHIR server configuration.

Reviewer: wow, interesting

Author: yeah - I left it there (originally from Robin)

1. Replace the following values with your service details:
    - `/path/to/db2jcc4.jar` : replace with the path to your driver jar. 
    - `<DB-HOSTNAME>`: the hostname of your Db2 service from the Service Credentials page
    - `<API-KEY>`: the API key value created in the previous section
    - `<TRUSTSTORE-PASSWORD>`: the password for your truststore
   
    - Note: When using an API Key, no username needs to be provided. This is because the API Key maps to a ServiceId, and that ServiceId is mapped to the Db2 user.

1. Run in your favorite terminal, and you should see no errors in the output. You should see output like: 

```
[jcc][10516][13709]Test Connection Successful.
```

### **Configuring IBM FHIR Server Datasource**

> The IBM FHIR server uses a proxy datasource mechanism, allowing new datasources to be added at runtime without requiring a (Liberty Profile) server restart. To configure a FHIR tenant datasource using an API-KEY, use the following template:

Reviewer: worth mentioning which file to put this in and where to find it?  or maybe just link to that section of the user's guide?

Author: I've now added.

Note, no username properties are given, because the authentication module only requires the API-KEY.

### **Configuring a BulkData's Liberty Datasource with API key**

The IBM FHIR Server Bulk Data modules utilize Java Batch (JSR-352) from the Liberty Profile feature - `batch-1.0`. The batch feature is configured to use Db2 as follows:

1. Create a Db2 user (e.g. FHIRBATCH)

> 1. and associate it with a ServiceId (no need to create an Administration user, a simple user has sufficient privileges). 

Reviewer: ```suggestion
1. Associate it with a ServiceId (no need to create an Administration user, a simple user has sufficient privileges). 
```

Author: grabbing in local commit

1. Using a valid API-KEY for the given ServiceId, configure a new datasource and the Java Batch persistence layer as follows:

``` xml
    <dataSource id="fhirbatchDS" jndiName="jdbc/fhirbatchDB">
        <jdbcDriver libraryRef="fhirSharedLib"/>
        <properties.db2.jcc
            serverName="dashdb-txn-flex-************.services.dal.bluemix.net"
            portNumber="50001"
            apiKey="<API-KEY>"
            securityMechanism="15"
            pluginName="IBMIAMauth"
            databaseName="BLUDB"
            currentSchema="JBATCH"
            driverType="4" sslConnection="true" sslTrustStoreLocation="resources/security/dbTruststore.jks" sslTrustStorePassword="<TRUSTSTORE-PASSWORD>"/>
    </dataSource>

    <batchPersistence jobStoreRef="BatchDatabaseStore" />
    <databaseStore id="BatchDatabaseStore" dataSourceRef="fhirbatchDS" schema="JBATCH" tablePrefix="" />
```

> - Note: The Java Batch is configured in `batchDs.xml` and included from the IBM FHIR Server's `server.xml` which is installed to the `{wlp}/usr/server/fhir-server`.

Reviewer: ```suggestion
- Note: The Java Batch is configured in `batchDs.xml` and included from the IBM FHIR Server's `server.xml` which is installed to `{wlp}/usr/server/fhir-server/`.
```

Author: I'm grabbing this in my local commit for simplicity
